CMSgov/easi-app
# Easi 1955/multiple email notifications revised

# EASI-1955

Initial draft of my work in allowing multiple email notifications to be sent when a GRT user takes actions. This PR only allows that for issuing and extending lifecycle IDs, but it also adds substantial test support through Cypress. I'd like some feedback on the general approach to both adding functionality and testing.

## Changes - Functionality
- I've added the `notifyMultipleRecipients` feature flag to LaunchDarkly - https://app.launchdarkly.com/default/dev/features/notifyMultipleRecipients/targeting. This flag is checked for a GRT user taking an action; if it's set to true, it allows them to notify arbitrary recipients, instead of the existing behavior of notifying the requester and CC'ing the Governance inbox.
- I've added an `EmailNotificationRecipients` input to the GQL schema, which contains a list of arbitrary recipients and options for notifying the Governance and/or Investment mailboxes. (Those addresses are set in configuration values) All actions which can send an email now take an optional `EmailNotificationRecipients` field, though right now it's only actually used when handling `IssueLifecycleIdInput` and `CreateSystemIntakeActionExtendLifecycleIdInput`. There's a corresponding Go struct for this GQL input in `pkg/models/email_address.go` called `EmailNotificationRecipients`.
- The bulk of the notification logic is implemented in methods in `pkg/email/` and functions in `pkg/services/`. The added methods in `email.Client` handle the logic of looping over the intended recipients and sending them emails, accumulating and tracking errors with the [go-multierror package](https://pkg.go.dev/github.com/hashicorp/go-multierror). The service functions now check the `notifyMultipleRecipients` feature flag; if it's set, they call the email client methods. The logic is kind of hard to view in the diff view, it's easier to understand if you pull the branch and look at it locally.

## Changes - Testing
- The email notification testing is in `cypress/integration/emailNotifications.spec.js`. Each test uses the Mailcatcher container's API to check what emails have or haven't been sent. 
- [Custom Cypress tasks](https://docs.cypress.io/api/commands/task) are defined in `cypress/plugins/index.js` for issuing/extending LCIDs by directly calling our GraphQL API. Once we complete [EASI-2019](https://jiraent.cms.gov/browse/EASI-2019), we should be able to remove these tasks and select recipients through the UI.
- Our LaunchDarkly client wrapper now accepts a third `FLAG_SOURCE`, `FILE`, which loads preset flag values from a file for testing purposes; see the [LaunchDarkly docs](https://docs.launchdarkly.com/sdk/features/flags-from-files) for more details. The code for configuring this is in `pkg/flags/client.go`; there's also a fair amount of configuration for plumbing `FLAGSOURCE=FILE` and `FLAGDATA_FILE` through to the appropriate containers.
- `cypress/fixtures/flagdata.json` was added, which has flag values for the email notifications. It sets `notifyMultipleRecipients` to true for the user `CYPT` and false for the user `CPYF`, which are used in the Cypress tests. It also disables the feature flags for connecting to CEDAR.

## How to test this change

1. Run the application locally.
2. Run `yarn cypress:open` to open the Cypress GUI.
3. Open the `emailNotifications.spec.js` spec to run the tests. (I've seen Cypress crash due to OOM issues when running all 8 tests; if you run into this, try using [`.only`](https://docs.cypress.io/guides/core-concepts/writing-and-organizing-tests#Excluding-and-Including-Tests) to only run tests for issuing LCIDs, or only the tests for extending LCIDs.
4. Open the Mailcatcher UI at localhost:1080 and see the emails that are sent as notifications.

## Known Issues
- Something in the email notification tests causes Cypress issues; there's the aforementioned OOM problems, and the tests take a long time to run in CI. I'd like to address whatever's causing that.

## Future Work
These will be addressed in a future PR.
- Obviously, there are still several other actions that can trigger emails; I need to implement notification functionality and testing for them.
- Once https://github.com/CMSgov/easi-app/pull/1695 is merged, I'll add the Investment mailbox to the email client config, as well as incorporating both it and the Governance mailbox in the Cypress tests.
- I'd like to see if there's a way to make the Cypress tests more table-driven, so the test code is less repetitious.
